File: fortran/angular_momentum.f90
module angular_momentum
   use iso_fortran_env, only: dp => real64
   use constants, only: num_particles, particle_mass

   implicit none (type, external)

contains

   subroutine compute_angular_momenta_in_z_direction(positions, momenta, angular_momenta)
      real(kind=dp), dimension(num_particles, 4), intent(in) :: &
         positions(:, :), momenta(:, :)

      real(kind=dp), dimension(num_particles), allocatable, intent(out) :: angular_momenta(:)

      integer :: i

      allocate(angular_momenta(1:num_particles))

      write(*, '("Computing angular momenta in the z direction")')

      do i = 1, num_particles
         ! Compute m * (x * v_y - y * v_x)
         angular_momenta(i) = particle_mass * (positions(i, 2) * momenta(i, 3) - positions(i, 3) * momenta(i, 2))
      end do

      write(*, '("Done computing angular momenta")')

   end subroutine compute_angular_momenta_in_z_direction

end module angular_momentum

File: rust/src/angular_momentum.rs
use crate::types::{Real, Vec4};

/// Compute the angular momenta (in the z direction) of a bunch of particles.
pub fn compute_angular_momenta_in_z_direction(
    positions: &Vec<Vec4>,
    momenta: &Vec<Vec4>,
) -> Vec<Real> {
    assert_eq!(positions.len(), momenta.len());

    positions
        .iter()
        .zip(momenta.iter())
        .map(|(u, v)| u[1] * v[2] - u[2] * v[1])
        .collect()
}
